File: curso_c_cpp/c/07_funciones/ejemplos/recursion.c
/*
 * Recursion: factorial y Fibonacci. Cada una tiene su caso base.
 */
#include <stdio.h>

long long factorial(int n) {
    if (n <= 1) return 1;                  // caso base
    return n * factorial(n - 1);           // caso recursivo
}

int fibonacci(int n) {
    if (n < 2) return n;                    // fib(0)=0, fib(1)=1
    return fibonacci(n - 1) + fibonacci(n - 2);
}

int main(void) {
    printf("factorial(5) = %lld\n", factorial(5));   // 120
    printf("fibonacci(10) = %d\n", fibonacci(10));    // 55
    return 0;
}

File: curso_java/07_metodos/ejemplos/Recursion.java
/*
 * Recursion: factorial y Fibonacci, cada una con su caso base.
 */
public class Recursion {

    static long factorial(int n) {
        if (n <= 1) return 1;
        return n * factorial(n - 1);
    }
    static int fibonacci(int n) {
        if (n < 2) return n;
        return fibonacci(n - 1) + fibonacci(n - 2);
    }

    public static void main(String[] args) {
        System.out.println("factorial(5) = " + factorial(5));    // 120
        System.out.println("fibonacci(10) = " + fibonacci(10));  // 55
    }
}
